Multi-threading implementation

fheinz opened the pull request:
Separate BLE and motor control into separate threads.

This seems to work more reliably than the single-threaded implementation, but I'm not very sure yet that it's free of synchronization issues.

fheinz committed:
Multi-threading implementation

Separate BLE and motor control into separate threads.

This seems to work more reliably. I don't think we're quite where we want in terms of synchronization, need to work out a way to make sure both gloves are on the same page about the pattern in a way that is more by design than by chance.

fheinz committed:
Merge branch 'matthewlai:main' into multitasking

matthewlai commented on firmware/main/main.ino:
```diff
@@ -325,31 +333,60 @@ bool EnsureBLEConnectionToServer() {
+}
+
+void motor_client_task(void *parameter) {
+  while (next_motor_cycle_start <= xTaskGetTickCount() ) vTaskDelay(10 / portTICK_PERIOD_MS);
```
This assumes the client is running behind the server right? If the client is running before the server, it wouldn't be able to catch up?

fheinz replied:
In the client, `next_motor_cycle_start` is expressed in ticks of the local clock, not of the server's. The adjustment happens un `ClientNotifyCallback`, on line 279: `next_motor_cycle_start = next_cycle_start + tick_drift;`

fheinz replied:
In other words, all this does is get the client to wait until it has a target time at which it needs to start vibrating for the first time. From then on, it just follows the lead of the server.

matthewlai replied:
If the drift ends up being negative (the client is behind), neither of the delays in this function would happen, the client will execute the pattern late, and finish the pattern still late by the same amount, and then start the next pattern late by the same amount?

matthewlai replied:
That's why in my crude implementation there was a "freshness" check, so that if it's running behind, rather than starting that pattern late and finish late, it waits till the next start instead. But there are definitely better ways of solving that.

matthewlai replied:
One potentially very fun way of doing this is to turn the whole thing into a state machine, with a pure function that deterministically tells us which motor should be active, if any, given the current time.

Something like:

```
int which_motor_on(TickType_t corrected_tick_count) {
  auto cycle_number = corrected_tick_count / 1330;
  auto tick_in_cycle = corrected_tick_count % 1330;
  // hash the cycle_number or something to determine pattern for this cycle number
  if (tick_in_cycle >= 664) {
    return -1; // we are in the rest between sequences.
  } else {
    auto period = tick_in_cycle / 166;
    auto tick_in_period = period % 166;
    if (tick_in_period < 100) {
        return pattern[period];
    } else {
        return -1; // we are in the rest between fingers.
    }
  }
}
```

Then we just need to sync time.

fheinz replied:
Maybe what you are missing is the fact that the time that the server advertises is **not** the start time for the current cycle, but for the **next** cycle, in its own absolute time. When the client gets this notification, it computes the tick drift and uses it to convert the start time to its own absolute time.

Right now I'm disregarding the latency of the notification, which may be a mistake, maybe I need to do a round-trip sync thing to more accurately measure the drift, but I'm leaving that for a future version in case we need it.

The first loop is only relevant at startup for the client, and basically stalls the motors from being operated until the client has received at least one notification from the server. Until then, `next_motor_cycle_start` is zero and hence smaller than the tick count.

The second loop in that function, on the other hand, sleeps for as many ticks as remain to the `next_motor_cycle_start`. This keeps it from drifting, no matter how the clocks start.

I guess I should add some comments to at least document the intent, which may be different from what I coded :-)

matthewlai replied:
Ah yes, in that case if the client clock drifts slower, it will skip a cycle. However, this is a bit fragile as it depends on the drift being smaller than the notification latency, otherwise the client can get into a state where it skips every other cycle.

Imagine an exaggerated case where a cycle is 1s for simplicity, the server clock is accurate, and the client clock is 10% slower (1.1s per cycle).

At time 0s, server starts cycle 0, and tells client that cycle 1 starts at 1s. Then at 1s, it tells client that cycle 2 starts at 2s, etc.

Client starts cycle 1 at 1s, but doesn't finish until 2.1s. By then, the client has received the third notification already, so wait till 3.0s to start cycle 3 (skipping cycle 2). Then it finishes at 4.1s, and have to skip cycle 4, and wait for cycle 5, etc.

I imagine this is probably fine as the notification latency is likely more than clock drift in one cycle in normal circumstances, though with preemptive multi-threading, task scheduling latency may be enough to tip things over. The state machine approach would be robust to this, as both client and server know exactly what they need to be doing at any given time, and as long as we have reasonably well-synced time, there should be no cycle skipping.

That said, happy to merge this if you are happy with sticking to this approach for now.

matthewlai replied:
In the case that the client is slower by N, and the latency L, the current approach would skip a cycle every L/N cycles, when the accumulated drift catches up with the notification latency.

fheinz replied:
Indeed, if the drift were calculated only once, the scenario you describe could happen. However: 

* The cycle is 1s. If the notification latency is larger than that, we have bigger problems. 
* I'm computing the drift on each notification, so the relative speed of the clocks should not result in error accumulation.

That said, I'm not entirely happy with that approach either: I expect notification latency to not be uniform, which will lead to the drift wobbling.

One way of solving this would be to have the client poll for the synchronization data, which could give us a two-way handshake on which we can run a proper time synchronization algorithm.

Something simpler would be to apply exponential dampening to the drift, which would make it more stable. 

In any case, the thing seems to work well enough on my boards. I'm wondering what would be a good way to measure how well synchronized the motor signals are, other than donning the gloves and going "yeah, good enough" or "that feels odd."

Unfortunately, all I can use an oscilloscope for is as a coaster for my drink, and that's frowned upon. Maybe you can help? 

In the meantime, merging this in would let me continue to fiddle with it, so that would be nice.

matthewlai replied:
Ah ok I didn't notice that the final delay has been taken out of the cycle. That should be fine then.

Yeah I'll merge this now.

Happy to help with the scoping when you want to lend me your boards for a bit, as I only have one board now and no battery.

fheinz committed:
Fix the client's cycle

The client was calling RunMotors() twice per cycle, because the motor routine was ending before we got a new start of cycle.

Fixed by waiting until the cycle's end after the motor routine is done.

Also made the notification logging less noisy.
